pkg/endpoint/policy_test.go

```diff
@@ -38,3 +53,163 @@ func (s *EndpointSuite) TestUpdateVisibilityPolicy(c *check.C) {
+			id.ID: id.LabelArray,
+		}, nil, wg)
+		wg.Wait()
+		return id
+	}
+
+	podID := addIdentity("pod")
+	repo.GetPolicyCache().LocalEndpointIdentityAdded(podID)
+
+	ep := Endpoint{
+		SecurityIdentity: podID,
+		policyGetter:     &mockPolicyGetter{repo},
+	}
+	ep.UpdateLogger(nil)
+
+	podSelectLabel := labels.ParseSelectLabel("pod")
+	egressSelectLabel := labels.ParseSelectLabel("peer")
+
+	// Create a rule that selects our pod, and allows all peer identities
+	egressDenyRule := &api.Rule{
+		EndpointSelector: api.NewESFromLabels(podSelectLabel),
+		EgressDeny: []api.EgressDenyRule{
+			{
+				EgressCommonRule: api.EgressCommonRule{
+					ToEndpoints: []api.EndpointSelector{
+						api.NewESFromLabels(egressSelectLabel),
+					},
+				},
+				ToPorts: []api.PortDenyRule{
+					{
+						Ports: []api.PortProtocol{
+							{
+								Port:     "80",
+								Protocol: "TCP",
+							},
+						},
+					},
+				},
+			},
+		},
+		Labels: labels.LabelArray{
+			labels.NewLabel(k8sConst.PolicyLabelName, "egressDenyRule", labels.LabelSourceAny),
+		},
+	}
+
+	_, _, err := repo.Add(*egressDenyRule, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Track all IDs we allocate so we can validate later that we never miss any
+	checkMutex := lock.Mutex{}
+	allocatedIDs := make(sets.Set[identity.NumericIdentity], testfactor)
+	done := false
+
+	// simulate ipcache churn: continuously allocate IDs and push them to the policy engine.
+	go func() {
+		for i := 0; i < testfactor; i++ {
+			if i%100 == 0 {
+				t.Log(i)
+			}
+			id := addIdentity("peer", fmt.Sprintf("peer%d", i))
+
+			// note: we could stop checking here and the last ID would be missing from allocatedIDs
+			// so we will have to handle the case where we select one more ID than is in allocatedIDs
+			checkMutex.Lock()
+			allocatedIDs.Insert(id.ID)
+			checkMutex.Unlock()
+
+		}
+		done = true
+	}()
+
+	// Continuously compute policy for the pod and ensure we never missed an incremental update.
+	for {
+		t.Log("Calculating policy...")
+		res, err := ep.regeneratePolicy()
+		assert.Nil(t, err)
+
+		// Sleep a random amount, so we accumulate some changes
+		// This does not slow down the test, since we always generate testFactor identities.
+		time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
+
+		// Now, check that all the expected entries are there
+		checkMutex.Lock()
+		t.Log("Checking policy...")
+
+		// Apply any pending incremental changes
+		// This mirrors the existing code, where we consume map changes
+		// while holding the endpoint lock
+		res.endpointPolicy.ConsumeMapChanges()
+		haveIDs := make(sets.Set[identity.NumericIdentity], testfactor)
+		for k := range res.endpointPolicy.PolicyMapState {
+			haveIDs.Insert(identity.NumericIdentity(k.Identity))
+		}
+
+		// It is okay if we have *more* IDs than allocatedIDs, since we may have propagated
+		// an ID change through the policy system but not yet added to the extra list we're
```

Comment:
> since we may have propagated an ID change through the policy system

Specifically where could that have happened? Mostly just curious so we can explain that in the comment as well because it's not obvious to me.

Reply:
In order to simulate a "churning" ipcache, the test only stops the identity allocation loop to write to the `allocatedIDs` set. This means that, depending on the interleaving, we may have fully plumbed-through one more ID than is written to the set. In other words `len(have) == len(want) + 1`. I could add more locking to prevent this, but then the test wouldn't be as faithful to the actual system.